File: move_rejected.py
"""Move rejected files (rating -1) and their same-stem siblings to a _rejected folder."""

import shutil
from pathlib import Path
from typing import List, Tuple

REJECTED_DIR_NAME = "_rejected"


def find_siblings(file: Path) -> List[Path]:
    """Same-stem files in the same directory (RAW/JPG pairs, .xmp sidecars)."""
    stem = file.stem.lower()
    return [
        c for c in file.parent.iterdir()
        if c.is_file() and c != file and c.stem.lower() == stem
    ]


def collect_move_set(rejected: List[Path]) -> List[Path]:
    """Rejected files plus their siblings, deduplicated, order-preserving."""
    seen = set()
    result = []
    for f in rejected:
        for path in [f, *find_siblings(f)]:
            if path not in seen:
                seen.add(path)
                result.append(path)
    return result


def move_to_rejected(files: List[Path], root: Path) -> Tuple[int, str]:
    """Move files into root/_rejected/, preserving relative subpaths.

    Returns (moved_count, error). Stops on the first error; error is "" on success.
    """
    moved = 0
    for f in files:
        try:
            rel = f.relative_to(root)
        except ValueError:
            rel = Path(f.name)
        dest = root / REJECTED_DIR_NAME / rel
        try:
            dest.parent.mkdir(parents=True, exist_ok=True)
            shutil.move(str(f), str(dest))
            moved += 1
        except OSError as e:
            return moved, f"{f.name}: {e}"
    return moved, ""

File: mac/RAWViewer/Sources/Sidecar/MoveRejected.swift
import Foundation

/// Moving rejected files into `_rejected/`. Spec 05 §10.
public enum MoveRejected {
    public static let directoryName = "_rejected"

    /// Every other file in the same directory with the same case-insensitive stem.
    public static func findSiblings(of file: URL) -> [URL] {
        let directory = file.deletingLastPathComponent()
        let stem = file.deletingPathExtension().lastPathComponent.lowercased()
        guard let entries = try? FileManager.default.contentsOfDirectory(
            at: directory, includingPropertiesForKeys: [.isDirectoryKey], options: []) else { return [] }
        return entries.filter { candidate in
            guard candidate.lastPathComponent != file.lastPathComponent else { return false }
            let isDirectory = (try? candidate.resourceValues(forKeys: [.isDirectoryKey]).isDirectory) ?? false
            guard !isDirectory else { return false }
            return candidate.deletingPathExtension().lastPathComponent.lowercased() == stem
        }
    }

    /// Each rejected file followed by its siblings, deduplicated, first-seen order preserved.
    public static func collectMoveSet(_ rejected: [URL]) -> [URL] {
        var seen = Set<String>()
        var result: [URL] = []
        for file in rejected {
            for candidate in [file] + findSiblings(of: file) {
                let key = candidate.standardizedFileURL.path
                if seen.insert(key).inserted { result.append(candidate) }
            }
        }
        return result
    }

    /// Moves files into `root/_rejected/<relative path>`, preserving subpaths.
    /// Stops at the first error. Deliberate deviation from the Python app: an existing
    /// destination gets a ` 2`, ` 3`… suffix before the extension instead of being overwritten.
    public static func moveToRejected(_ files: [URL], root: URL) -> (moved: Int, error: String?) {
        let fm = FileManager.default
        let rootPath = root.standardizedFileURL.path
        let destinationRoot = root.appendingPathComponent(directoryName, isDirectory: true)
        var moved = 0

        for file in files {
            let source = file.standardizedFileURL
            let relative: String
            if source.path.hasPrefix(rootPath + "/") {
                relative = String(source.path.dropFirst(rootPath.count + 1))
            } else {
                relative = source.lastPathComponent
            }
            let destination = uniqueDestination(destinationRoot.appendingPathComponent(relative))
            do {
                try fm.createDirectory(at: destination.deletingLastPathComponent(),
                                       withIntermediateDirectories: true)
                try fm.moveItem(at: source, to: destination)
                moved += 1
            } catch {
                return (moved, "\(source.lastPathComponent): \(error.localizedDescription)")
            }
        }
        return (moved, nil)
    }

    static func uniqueDestination(_ url: URL) -> URL {
        let fm = FileManager.default
        guard fm.fileExists(atPath: url.path) else { return url }
        let directory = url.deletingLastPathComponent()
        let ext = url.pathExtension
        let stem = url.deletingPathExtension().lastPathComponent
        var suffix = 2
        while true {
            let name = ext.isEmpty ? "\(stem) \(suffix)" : "\(stem) \(suffix).\(ext)"
            let candidate = directory.appendingPathComponent(name)
            if !fm.fileExists(atPath: candidate.path) { return candidate }
            suffix += 1
        }
    }
}
